client-course-schedulizer/src/readCSV.ts

```diff
@@ -0,0 +1,428 @@
+  // eslint-disable-next-line prefer-const
+  const schedule: di.Schedule = {
+    sections: [],
+  };
+
+  // Get data and fields from the CSV
+  const { data, meta } = objects;
+  const { fields } = meta;
+
+  // From the CSV fields, take the ones which we recognize
+  // eslint-disable-next-line prefer-const
+  let usableFields: string[] = [];
+  let field;
+  let startTimeIdx = -1;
+  let startTimeStrIdx = -1;
+  let instructorIdx = -1;
+  let instructorsIdx = -1;
+  let prefixIdx = -1;
+  let prefixesIdx = -1;
+  if (fields) {
+    for (let i = 0; i < fields.length; i += 1) {
+      field = fields[i];
+      if (validFields.includes(field)) {
+        usableFields.push(field);
+        if (field === "startTime") {
+          startTimeIdx = usableFields.length - 1;
+        } else if (field === "startTimeStr") {
+          startTimeStrIdx = usableFields.length - 1;
+        } else if (field === "instructor") {
+          instructorIdx = usableFields.length - 1;
+        } else if (field === "instructors") {
+          instructorsIdx = usableFields.length - 1;
+        } else if (field === "prefix") {
+          prefixIdx = usableFields.length - 1;
+        } else if (field === "prefixes") {
+          prefixesIdx = usableFields.length - 1;
+        }
+      }
+    }
+  }
+
+  // If both "startTime" and "startTimeStr" are present, ignore "startTime"
+  // This is because Pruim's data has a timestamp in "startTime" (not our timezone)
+  if (startTimeIdx !== -1 && startTimeStrIdx !== -1) {
+    usableFields.splice(startTimeIdx, 1);
+  }
+  // If both "instructor" and "instructors" are present, ignore "instructor"
+  if (instructorIdx !== -1 && instructorsIdx !== -1) {
+    usableFields.splice(instructorIdx, 1);
+  }
+  // If both "prefix" and "prefixes" are present, ignore "prefix"
+  if (prefixIdx !== -1 && prefixesIdx !== -1) {
+    usableFields.splice(prefixIdx, 1);
+  }
+
+  // Parse each row of the CSV as an object
+  let object;
+  let value: string;
+  let regMatch: RegExpMatchArray | null;
+  let roomParts: string[];
+  let names: string[];
+  let nameParts: string[];
+  for (let i = 0; i < data.length; i += 1) {
+    object = data[i];
+
+    // Reset defaults
+    name = "";
+    prefixes = [];
+    number = "";
+    letter = "";
+    studentHours = 0;
+    facultyHours = 0;
+    hourPart = "8";
+    numHourPart = 8;
+    minPart = "00";
+    ampm = "AM";
+    startTime = "8:00 AM";
+    duration = 50;
+    building = "";
+    roomNumber = "";
+    roomCapacity = 30;
+    year = new Date().getFullYear();
+    term = di.Term.Fall;
+    half = di.Half.Full;
+    days = [di.Day.Monday, di.Day.Wednesday, di.Day.Friday];
+    globalMax = 30;
+    localMax = 30;
+    anticipatedSize = 30;
+    instructors = [];
+    comments = "";
+
+    // Iterate through the fields of the CSV, and parse their values for this object
+    for (let j = 0; j < usableFields.length; j += 1) {
+      field = usableFields[j];
+      if (field in object) {
```

Comment:
Since `usableFields` is taken from the parsed object, wouldn't `field` always be in `object`?

Reply:
I think it would probably be cleaner here to do `usableFields.forEach((field) => {... loop logic ...})`

Reply:
a413c2b